apps/glusterfs/pendingop_entry.go

```diff
@@ -0,0 +1,265 @@
+//
+// Copyright (c) 2018 The heketi Authors
+//
+// This file is licensed to you under your choice of the GNU Lesser
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), or the GNU General Public License, version 2 (GPLv2), in all
+// cases as published by the Free Software Foundation.
+//
+
+package glusterfs
+
+import (
+	"bytes"
+	"encoding/gob"
+	"time"
+
+	"github.com/boltdb/bolt"
+	wdb "github.com/heketi/heketi/pkg/db"
+	"github.com/heketi/heketi/pkg/utils"
+	"github.com/lpabon/godbc"
+)
+
+const (
+	NEW_ID                    = ""
+	BOLTDB_BUCKET_PENDING_OPS = "PENDING_OPERATIONS"
+	DB_HAS_PENDING_OPS        = "PENDING_OPERATIONS"
+)
+
+var (
+	// support unit test dep. injection for custom timestemps
+	operationTimestamp = func() int64 { return time.Now().Unix() }
+)
+
+// PendingOperationEntry tracks pending operations within the Heketi db.
+type PendingOperationEntry struct {
+	PendingOperation
+}
+
+// PendingOperationList returns the IDs of all pending operation entries
+// currently in the Heketi db.
+func PendingOperationList(tx *bolt.Tx) ([]string, error) {
+	list := EntryKeys(tx, BOLTDB_BUCKET_PENDING_OPS)
+	if list == nil {
+		return nil, ErrAccessList
+	}
+	return list, nil
+}
+
+// HasPendingOperations returns true if the db contains one or more pending
+// operation entries. If the db cannot be read the function panics.
+func HasPendingOperations(db wdb.RODB) bool {
+	var pending bool
+	if err := db.View(func(tx *bolt.Tx) error {
+		l, err := PendingOperationList(tx)
+		if err != nil {
+			return err
+		}
+		pending = (len(l) > 0)
+		return nil
+	}); err != nil {
+		panic(err)
+	}
+	return pending
+}
+
+// BucketName returns the name of the db bucket for pending operation entries.
+func (p *PendingOperationEntry) BucketName() string {
+	return BOLTDB_BUCKET_PENDING_OPS
+}
+
+// NewPendingOperationEntry returns a newly constructed pending operation entry.
+// If id is a non-zero-length string then that value will be used as the ID of
+// the object. Otherwise pass NEW_ID to have a new uuid be automatically allocated
+// for the new object.
+func NewPendingOperationEntry(id string) *PendingOperationEntry {
+	if id == NEW_ID {
+		id = utils.GenUUID()
+	}
+	entry := &PendingOperationEntry{PendingOperation{
+		PendingItem: PendingItem{id},
+		Timestamp:   operationTimestamp(),
+		Actions:     []PendingOperationAction{},
+	}}
+	return entry
+}
+
+// NewPendingOperationEntryFromId fetches an existing pending operation entry
+// from the heketi db based on the provided id.
+func NewPendingOperationEntryFromId(tx *bolt.Tx, id string) (
+	*PendingOperationEntry, error) {
+	godbc.Require(tx != nil)
+	godbc.Require(id != "")
+
+	entry := &PendingOperationEntry{}
+	err := EntryLoad(tx, entry, id)
+	if err != nil {
+		return nil, err
+	}
+
+	if entry.Actions == nil {
+		entry.Actions = []PendingOperationAction{}
```

Comment:
is this required?

Reply:
Not sure really. This pattern is fairly frequent in other *_entry.go files so I figured better safe than sorry in that regard and copied that pattern.

Reply:
makes sense..